### ISSUE TITLE
Idea: Use PapaParse for handling CSV files

### ISSUE THREAD

**Issue by bean5** [SheetJS/js-xlsx#1278]
Since noticing that CSV files are modified upon being read, I realized that perhaps I should just use another library for CSV files. However, I didn't like that idea because they really are just another form of spreadsheet.

For now, I am using PapaParse, but ideally js-xlsx would use it so I can just rely on js-xlsx and let it do its magic!

**Comment by SheetJSDev** [comment]
There were quite a few issues raised / commented upon, so it's better to just put the related responses in one message.

tl;dr: Excel "modifies" CSV files as well.  SheetJS tries to mimic Excel behavior.  Other libraries try to follow RFC4180, blissfully ignoring Excel's behavior.

The Excel formats, as well as the Lotus formats, Symbolic Link, and other formats, separate "content" from "presentation".  This is a great decision because it lets you show numbers with fewer decimal places while still retaining the original values for future calculations.  For each cell, Excel stores the number format as well as the actual underlying IEEE754 number (but not the final formatted string!).  In the library, we represent the "content" as the `.v` attribute of a cell, the "number format" as the `.z` attribute, and the "presentation" as the `.w` attribute (these are single-character because it is common to have to serialize the workbook object as JSON -- like when working across a web worker -- so each additional character adds up when dealing with millions of cells)

CSV mixes both.  Even though there's an alleged specification RFC4180, it leaves the "Value" part completely implementation-dependent.  There's no standard for what constitutes a date, what constitutes a number, and in fact there's no concept of data types.  Nothing says that the string "2018-09-19" is a Date.  It is the end application that has to decide what to do.

Excel has many special behaviors that are not standard according to RFC4180.  In addition to features like automatic value interpretation, there is the `sep` directive to specify the field separator:

```
sep=;
1;2;3
4;5;6
```

Related to the original question, if you try opening CSV dates in Excel, it automatically falls back to number code 14 (the default d/m/yy format)!  For example, consider the CSV:

```
1,2,3
a,b,c
2018-09-19
```

Save this to a file and open in Excel, and you will see that Excel throws away the original format!

<img width="204" alt="" src="https://user-images.githubusercontent.com/6070939/45792492-eed85180-bc5b-11e8-91c1-0ce216a43848.png">

Reading the data with `cellDates: true` will preserve the original date and use the default date format according to Excel:

```js
XLSX.read("1,2,3\na,b,c\n2018-09-19", {type:"binary", cellDates:true, cellNF: true}).Sheets.Sheet1
{ A1: { t: 'n', w: '1', v: 1 },
  B1: { t: 'n', w: '2', v: 2 },
  C1: { t: 'n', w: '3', v: 3 },
  A2: { t: 's', v: 'a' },
  B2: { t: 's', v: 'b' },
  C2: { t: 's', v: 'c' },
  A3: { z: 'm/d/yy', t: 'd', v: 2018-09-19T00:00:00.000Z, w: '9/19/18' },
  '!ref': 'A1:C3' }
```

In summary, since RFC4180 is easy to approximate, pretty much all libraries go that route; since we're trying to replicate Excel's behavior as closely as possible, we have to do something different.

You did happen to reveal a bug: Excel will show hours/minutes if it thinks you specified a date time.  For example, the CSV value `2018-09-19 01:23:45` is rendered as `9/19/18 1:23`.  Even so, it's still "modifying" the CSV.

**Comment by bean5** [comment]
When you say Excel modifies CSVs do you mean in it's underlying data or what is surfaced to the user you opening a CSV?

I didn't know that this project aimed to mimic Excel's handling of files. I suppose there are user cases for that. My greatest surprise is that Excel modifies CSVs. Thanks for letting me know. Perhaps you could update your projects READ E to be more clear that CSVs are handled as if opened by Excel. The README mentioned compatibility, but that didn't last me to realize this before going the ticket.
